test/functional/readpreference.test.js

```diff
@@ -759,4 +759,34 @@ describe('ReadPreference', function() {
         });
     })
   });
+
+  it('should use session readPreference instead of client readPreference', {
+    metadata: { requires: { unifiedTopology: true, topology: ['single', 'replicaset'] } },
+    test: function(done) {
+      const configuration = this.configuration;
+      const client = this.configuration.newClient(configuration.writeConcernMax(), {
+        readPreference: 'primaryPreferred'
+      });
+
+      client.connect((err, client) => {
+        this.defer(() => {
+          client.close();
+        });
+        expect(err).to.not.exist;
+        expect(client).to.exist;
+        const session = client.startSession({
+          defaultTransactionOptions: { readPreference: 'secondary' },
+          causalConsistency: true
+        });
+
+        session.startTransaction();
+        const result = ReadPreference.resolve(client, { session: session });
+        expect(result).to.exist;
+        expect(result.mode).to.deep.equal('secondary');
+        session.abortTransaction();
+
+        session.endSession(undefined, done);
```

Comment:
I strongly recommend using beforeEach/afterEach for setup and clean up to ensure that the clean up step gets executed even if some assertions inside the test fail; for example, if your expect fails at L785, it's going to result in a thrown error and hence interrupt the flow of the code; 

to limit the impact of this change, feel free to use an extra describe block around this 'it' and declare a `let client` and `let session` in that scope; the beforeEach can take a done argument, so you should be able to write something like `beforeEach(done => { client = < config setup code >; client.connect(err => done(err); })` then inside the `it` block, have `session = client.startSession(...)` and so forth until L785; then define `afterEach(done => { if (session) { session.abortTransaction(() => session.endSession(() => client.close(() => done())))} else { client.close(() => done() } })`

Reply:
Thanks for this. That cleared up a lot of why we use the `beforeEach` and `afterEach`.